### PR TITLE
Support damage entity with damage-causes in MC 1.20.4+

### PR DESCRIPTION
**Pull request by 0XPYEX0**
### Description

As title
---
**Target Minecraft Versions:** 1.20.4+ 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

**Review by sovdeeth — CHANGES_REQUESTED** [review]
tests failing

**Review by cheeezburga — COMMENTED** [review]
Looks pretty good, just some things here and there. Also, would love to see the description and examples updated.

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 1]
```diff
@@ -58,18 +61,23 @@ public class EffHealth extends Effect {
...
+		if (matchedPattern == 0 && exprs[2] != null && !canSetDamageCause)
+			Skript.warning("The fake damage cause extension of this effect has no functionality in your server, it need MC 1.20.4+");
 
 		this.damageables = exprs[0];
 		this.isHealing = matchedPattern >= 1;
 		this.isRepairing = matchedPattern == 2;
 		this.amount = (Expression<Number>) exprs[1];
+		if (exprs.length > 2) {
+			this.exprCause = (Expression<EntityDamageEvent.DamageCause>) exprs[2];
+		}
```
[the remark this answers is not in the document]
Comment (0XPYEX0) [reply src/main/java/ch/njol/skript/effects/EffHealth.java:80]: IndexOutBounds error if we don't check array length😢

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 2]
```diff
@@ -34,42 +34,49 @@
...
+		if (matchedPattern == 0 && exprs[2] != null && !canSetDamageCause)
+			Skript.warning("Using the fake cause extension in effect 'damage' requires Spigot 1.20.4+");
 
 		this.damageables = exprs[0];
 		this.isHealing = matchedPattern >= 1;
 		this.isRepairing = matchedPattern == 2;
 		this.amount = (Expression<Number>) exprs[1];
+		if (exprs.length > 2) {
+			this.exprCause = (Expression<DamageCause>) exprs[2];
+		}
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:79]:
```suggestion
		if (exprs.length > 2)
			this.exprCause = (Expression<DamageCause>) exprs[2];
```

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 3]
```diff
@@ -126,14 +136,12 @@ protected void execute(Event event) {
 
 	@Override
 	public String toString(@Nullable Event event, boolean debug) {
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:138]: Needs to be updated to reflect a damage source if provided

File: src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java [hunk 4]
```diff
@@ -142,4 +143,34 @@ public static void setDamageCause(Damageable e, DamageCause cause) {
 		}
 	}
 
+	@NotNull
+	public static DamageSource getDamageSourceFromCause(DamageCause cause) {
```
[the remark this answers is not in the document]
Comment (0XPYEX0) [reply src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java:147]: But why don't you make `@Override` annotation in line? 😂

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 5]
```diff
@@ -1,19 +1,19 @@
 /**
```
[the remark this answers is not in the document]
Comment (0XPYEX0) [reply src/main/java/ch/njol/skript/effects/EffHealth.java:1]: I think I don't have the permission to remove this. It should be removed by a SkriptLang team member

[the remark this answers is not in the document]
Comment (Fusezion) [reply src/main/java/ch/njol/skript/effects/EffHealth.java:1]: Licenses are being removed, this is approved by the team due to it being bloat, I sadly couldn't find a good image in discord to prove it. But any recent pr probably you can see them removing it.

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 6]
```diff
@@ -109,14 +111,19 @@ protected void execute(Event event) {
...
+			} else if (obj instanceof Damageable damageable) {
 				if (this.amount == null) {
 					HealthUtils.heal(damageable, HealthUtils.getMaxHealth(damageable));
 				} else if (isHealing) {
 					HealthUtils.heal(damageable, amount);
 				} else {
+					if (canSetDamageCause && exprCause != null) {
+						DamageCause cause = exprCause.getSingle(event);
+						if (cause != null) {
+							damageable.damage(amount * 2, HealthUtils.getDamageSourceFromCause(cause));
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffHealth.java]: This isn't consistent with the other methods - you should make a method in HealthUtils for this, which would be mostly the same as `damage(damageable, amount)`, but with an additional `cause` parameter

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 7]
```diff
@@ -126,15 +133,15 @@ protected void execute(Event event) {
...
 		if (isRepairing) {
 			prefix = "repair ";
 		} else if (isHealing) {
 			prefix = "heal ";
 		}
-
-		return prefix + damageables.toString(event, debug) + (amount != null ? " by " + amount.toString(event, debug) : "");
+		return prefix + damageables.toString(event, debug)
+				   + (amount != null ? " by " + amount.toString(event, debug) : "")
+				   + (exprCause != null && event != null ? "with damage cause " + exprCause.getSingle(event) : "");
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:145]:
```suggestion
				   + (exprCause != null && event != null ? " with damage cause " + exprCause.getSingle(event) : "");
```

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 8]
```diff
@@ -34,42 +34,46 @@
 import ch.njol.util.Math2;
 import org.bukkit.entity.Damageable;
 import org.bukkit.event.Event;
+import org.bukkit.event.entity.EntityDamageEvent.DamageCause;
 import org.bukkit.inventory.ItemStack;
 import org.eclipse.jdt.annotation.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
 
 @Name("Damage/Heal/Repair")
 @Description("Damage/Heal/Repair an entity, or item.")
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:44]:
```suggestion
@Description("Damage, heal, or repair an entity or item.")
```
As well as this, the description should be updated a little bit to explain the damage cause stuff.

File: src/main/java/ch/njol/skript/effects/EffHealth.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (0XPYEX0) [reply src/main/java/ch/njol/skript/effects/EffHealth.java:1]: All right, I've got it.

**Review by cheeezburga — CHANGES_REQUESTED** [review]

**Comment by cheeezburga** [comment]
> Maybe I should move `HealthUtils.damage()` to `DamageUtils.damage()` ? @cheeezburga

I don't think so. All the methods that modify the health of a damageable should be left in HealthUtils.

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 9]
```diff
@@ -34,42 +17,46 @@
...
+	private final boolean canSetDamageCause = Skript.classExists("org.bukkit.damage.DamageSource");
 	private Expression<?> damageables;
-	@Nullable
-	private Expression<Number> amount;
+	private @UnknownNullability Expression<Number> amount;
 	private boolean isHealing, isRepairing;
+	private @UnknownNullability Expression<DamageCause> exprCause = null;
 
 	@SuppressWarnings("unchecked")
 	@Override
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:49]:
```suggestion
	@Override
	@SuppressWarnings("unchecked")
```

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 10]
```diff
@@ -34,42 +17,46 @@
...
 @Name("Damage/Heal/Repair")
-@Description("Damage/Heal/Repair an entity, or item.")
+@Description("Damage, heal, or repair an entity or item.")
 @Examples({
 	"damage player by 5 hearts",
+	"damage player by 3 hearts with fake cause fall",
 	"heal the player",
 	"repair tool of player"
 })
 @Since("1.0")
```
Comment (Fusezion) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:33]:
```suggestion
})
@Since("1.0, INSERT VERSION (fake damage cause)")
```

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 11]
```diff
@@ -34,42 +17,46 @@
...
 import org.bukkit.event.Event;
+import org.bukkit.event.entity.EntityDamageEvent.DamageCause;
 import org.bukkit.inventory.ItemStack;
-import org.eclipse.jdt.annotation.Nullable;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
 
 @Name("Damage/Heal/Repair")
-@Description("Damage/Heal/Repair an entity, or item.")
+@Description("Damage, heal, or repair an entity or item.")
```
Comment (Fusezion) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:26]: Description should be updated to include mention of fake damage

File: src/main/java/ch/njol/skript/effects/EffHealth.java (hunk 10 above)
Comment (Fusezion) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:33]: While I'm at this let's see about adding a `@RequiredPlugins` annotation pointing towards 1.20.4+

File: src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java [hunk 12]
```diff
@@ -31,13 +15,22 @@
 import org.bukkit.event.entity.EntityDamageEvent.DamageCause;
 import org.jetbrains.annotations.Nullable;
 
-import java.lang.reflect.Constructor;
-import java.lang.reflect.InvocationTargetException;
-
 public class HealthUtils {
+	@Nullable
+	private static final Constructor<EntityDamageEvent> OLD_DAMAGE_EVENT_CONSTRUCTOR;
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java:20]:
```suggestion
	
	private static final @Nullable Constructor<EntityDamageEvent> OLD_DAMAGE_EVENT_CONSTRUCTOR;
```

File: src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java [hunk 13]
```diff
@@ -31,13 +15,22 @@
...
+
+	static {
+		Constructor<EntityDamageEvent> constructor = null;
+		try {
+			constructor = EntityDamageEvent.class.getConstructor(Damageable.class, DamageCause.class, double.class);
+		} catch (NoSuchMethodException ignored) {
+		}
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java:27]:
```suggestion
		} catch (NoSuchMethodException ignored) {}
```

File: src/main/java/ch/njol/skript/effects/EffHealth.java (hunk 10 above)
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:33]:
```suggestion
@Since("1.0")
@RequiredPlugins("Spigot 1.20.4+ (fake damage cause)")
```

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 14]
```diff
@@ -34,42 +17,46 @@
...
+	private @UnknownNullability Expression<DamageCause> exprCause = null;
 
 	@SuppressWarnings("unchecked")
 	@Override
 	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
-		if (matchedPattern == 0 && exprs[2] != null)
-			Skript.warning("The fake damage cause extension of this effect has no functionality, " +
-				"and will be removed in the future");
+		if (matchedPattern == 0 && exprs[2] != null && !canSetDamageCause)
+			Skript.warning("Using the fake cause extension in effect 'damage' requires Spigot 1.20.4+");
```
Comment (Fusezion) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:52]: Do we want to throw an error or the warning? I don't think I'd like code running if there's invalid usage going through. The warning has existed long enough prior that there should be near zero percent concern of breaking change and this would hopefully force version checking or similar for people wanting to support older versions of mc.

**Review by Fusezion — CHANGES_REQUESTED** [review]
Just some mile issues with this, outside of missing any addition test for `with fake damage cause` so we should look into that as well.

**Review by cheeezburga — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/effects/EffHealth.java (hunk 14 above)
Comment (cheeezburga) [reply src/main/java/ch/njol/skript/effects/EffHealth.java:52]: Yeah I think erroring out here is better. Warning works, but it might be misleading to some users, and it'd make the logic later on easier too.

File: src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java (hunk 13 above)
Comment (0XPYEX0) [reply src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java:27]: Sorted by IDEA

File: src/main/java/ch/njol/skript/bukkitutil/DamageUtils.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
+package ch.njol.skript.bukkitutil;
+
+import org.bukkit.damage.DamageSource;
+import org.bukkit.damage.DamageType;
+import org.bukkit.event.entity.EntityDamageEvent.DamageCause;
+import org.jetbrains.annotations.NotNull;
+
+public class DamageUtils {
+	@NotNull
+	@SuppressWarnings("all")
```
[the remark this answers is not in the document]
Comment (0XPYEX0) [reply src/main/java/ch/njol/skript/bukkitutil/DamageUtils.java:9]: Actually IDK what is it
image

[the remark this answers is not in the document]
Comment (cheeezburga) [reply src/main/java/ch/njol/skript/bukkitutil/DamageUtils.java:9]: I think its `@SuppressWarnings("UnstableApiUsage")`

File: src/main/java/ch/njol/skript/effects/EffHealth.java (hunk 14 above)
Comment (0XPYEX0) [reply src/main/java/ch/njol/skript/effects/EffHealth.java:54]: I don't think we should error the code. We also can use `damage` effect even if it doesn't support for `damage cause`

File: src/main/java/ch/njol/skript/bukkitutil/DamageUtils.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
+package ch.njol.skript.bukkitutil;
+
+import org.bukkit.damage.DamageSource;
+import org.bukkit.damage.DamageType;
+import org.bukkit.event.entity.EntityDamageEvent.DamageCause;
+import org.jetbrains.annotations.NotNull;
+
+public class DamageUtils {
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/bukkitutil/DamageUtils.java:8]:
```suggestion
public class DamageUtils {

```

File: src/main/java/ch/njol/skript/bukkitutil/DamageUtils.java [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+			case FLY_INTO_WALL -> DamageType.FLY_INTO_WALL;
+			case HOT_FLOOR -> DamageType.HOT_FLOOR;
+			case CAMPFIRE -> DamageType.CAMPFIRE;
+			case CRAMMING -> DamageType.CRAMMING;
+			case DRYOUT -> DamageType.DRY_OUT;
+			case FREEZE -> DamageType.FREEZE;
+			case SONIC_BOOM -> DamageType.SONIC_BOOM;
+			default -> DamageType.GENERIC;
+		}).build();
+	}
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/bukkitutil/DamageUtils.java:38]:
```suggestion
	}
	
```

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 18]
```diff
@@ -34,42 +18,52 @@
...
 import org.bukkit.inventory.ItemStack;
-import org.eclipse.jdt.annotation.Nullable;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
 
 @Name("Damage/Heal/Repair")
-@Description("Damage/Heal/Repair an entity, or item.")
+@Description({
+	"Damage, heal, or repair an entity or item.",
+	"We can use damage entity with a fake damage cause in Spigot 1.20.4+"
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:29]:
```suggestion
	"Servers running Spigot 1.20.4+ can optionally choose to specify a fake damage cause."
```

File: src/main/java/ch/njol/skript/effects/EffHealth.java (hunk 10 above)
Comment (cheeezburga) [reply src/main/java/ch/njol/skript/effects/EffHealth.java:37]: This wasn't resolved

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 19]
```diff
@@ -34,42 +18,52 @@
...
-
 	static {
 		Skript.registerEffect(EffHealth.class,
-			"damage %livingentities/itemtypes/slots% by %number% [heart[s]] [with fake cause %-damagecause%]",
+			"damage %livingentities/itemtypes/slots% by %number% [heart[s]] [with [fake] [damage] cause %-damagecause%]",
 			"heal %livingentities% [by %-number% [heart[s]]]",
 			"repair %itemtypes/slots% [by %-number%]");
 	}
 
+	private final boolean canSetDamageCause = Skript.classExists("org.bukkit.damage.DamageSource");
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:47]: I think this can be made a static constant, and moved above the static block
```suggestion
	private static final boolean HAS_DAMAGE_SOURCE = Skript.classExists("org.bukkit.damage.DamageSource");
```

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 20]
```diff
@@ -109,14 +101,19 @@ protected void execute(Event event) {
...
 					HealthUtils.heal(damageable, amount);
 				} else {
+					if (canSetDamageCause && exprCause != null) {
+						DamageCause cause = exprCause.getSingle(event);
+						if (cause != null) {
+							HealthUtils.damage(damageable, amount, DamageUtils.getDamageSourceFromCause(cause));
+							return;
+						}
+					}
 					HealthUtils.damage(damageable, amount);
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:117]:
```suggestion
					DamageCause cause = exprCause == null ? null : exprCause.getSingle(event);
					if (cause != null) {
						HealthUtils.damage(damageable, amount, DamageUtils.getDamageSourceFromCause(cause));
						return;
					}
					HealthUtils.damage(damageable, amount);
```

**Review by cheeezburga — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java [hunk 21]
```diff
@@ -46,18 +37,21 @@ public static double getHealth(Damageable e) {
...
 	/**
 	 * Set the health of an entity
-	 * @param e Entity to set health for
+	 *
+	 * @param e      Entity to set health for
 	 * @param health The amount of hearts to set
 	 */
 	public static void setHealth(Damageable e, double health) {
 		e.setHealth(Math2.fit(0, health, getMaxHealth(e)) * 2);
+		//
```
Comment (Fusezion) [inline src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java:49]:
```suggestion
```

File: src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java [hunk 22]
```diff
@@ -91,8 +87,17 @@ public static void damage(Damageable e, double d) {
 		e.damage(d * 2);
 	}
 
+	public static void damage(Damageable e, double d, DamageSource cause) {
+		if (d < 0) {
```
Comment (Fusezion) [inline src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java:91]: Didn't realize this but we need proper naming of these values `e`, `d` don't explain anything `e` could be `event`, `d` could be `damageable`

Since we're doing it here we might as well apply these to everything in the class rather than just here

**Review by Fusezion — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 23]
```diff
@@ -34,42 +18,52 @@
...
+	"Damage, heal, or repair an entity or item.",
+	"Servers running Spigot 1.20.4+ can optionally choose to specify a fake damage cause."
+})
 @Examples({
 	"damage player by 5 hearts",
+	"damage player by 3 hearts with fake cause fall",
 	"heal the player",
 	"repair tool of player"
 })
 @Since("1.0")
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:37]:
```suggestion
@Since("1.0, INSERT VERSION (damage cause)")
```

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 24]
```diff
@@ -34,42 +18,52 @@
...
-
 	static {
 		Skript.registerEffect(EffHealth.class,
-			"damage %livingentities/itemtypes/slots% by %number% [heart[s]] [with fake cause %-damagecause%]",
+			"damage %livingentities/itemtypes/slots% by %number% [heart[s]] [with [fake] [damage] cause %-damagecause%]",
 			"heal %livingentities% [by %-number% [heart[s]]]",
 			"repair %itemtypes/slots% [by %-number%]");
 	}
 
+	private static final boolean canSetDamageCause = Skript.classExists("org.bukkit.damage.DamageSource");
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:47]:
```suggestion
	private static final boolean SUPPORTS_DAMAGE_SOURE = Skript.classExists("org.bukkit.damage.DamageSource");
```

File: src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java [hunk 25]
```diff
@@ -31,115 +15,123 @@
...
 	/**
 	 * Get the health of an entity
-	 * @param e Entity to get health from
+	 *
+	 * @param entity Entity to get health from
 	 * @return The amount of hearts the entity has left
 	 */
-	public static double getHealth(Damageable e) {
-		if (e.isDead())
+	public static double getHealth(Damageable entity) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java:35]: this should be damageable instead of entity, same in the rest of the class

**Review by Efnilite — CHANGES_REQUESTED** [review]

**Comment by Efnilite** [comment]
needs test

File: src/main/java/ch/njol/skript/effects/EffHealth.java [hunk 26]
```diff
@@ -34,42 +18,52 @@
...
+	"Servers running Spigot 1.20.4+ can optionally choose to specify a fake damage cause."
+})
 @Examples({
 	"damage player by 5 hearts",
+	"damage player by 3 hearts with fake cause fall",
 	"heal the player",
 	"repair tool of player"
 })
-@Since("1.0")
+@Since("1.0, 2.10 (damage cause)")
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffHealth.java:37]: 2.10 should be `INSERT VERSION`, it'll be replaced by the proper version when it gets released

**Review by Efnilite — APPROVED** [review]
nice

**Review by sovdeeth — CHANGES_REQUESTED** [review]
needs some tests but looks good

**Comment by 0XPYEX0** [comment]
What should I do for test？😂I don know it actually

**Comment by Efnilite** [comment]
> What should I do for test？😂I don know it actually

you should spawn an entity and assert that a bunch of normal and edge cases work as you expect. you can check out examples

File: src/test/skript/tests/syntaxes/effects/EffHealth.sk [hunk 27]
```diff
@@ -20,4 +20,7 @@ test "health effect":
 	assert health of {_m} is 1.5 with "heal cow failed"
 	heal {_m} by 99
 	assert health of {_m} is 5 with "heal cow failed"
+	wait 1 tick
+	damage {_m} by 2 hearts with damage cause fall
+	assert health of {_m} is 3 with "damage cow failed"
```
[the remark this answers is not in the document]
Comment (0XPYEX0) [reply src/test/skript/tests/syntaxes/effects/EffHealth.sk]: We must wait, because:
image
image
It just marked the entity to be damaged, the damage will be delayed

[the remark this answers is not in the document]
Comment (0XPYEX0) [reply src/test/skript/tests/syntaxes/effects/EffHealth.sk]: And there is something wrong with `parse if` section:
image
image

I can only upload the `parse if` version and wait for it to be fixed

**Comment by 0XPYEX0** [comment]
idk how to fix the `mark` problem

File: src/test/skript/tests/syntaxes/effects/EffHealth.sk (hunk 27 above)
[the remark this answers is not in the document]
Comment (sovdeeth) [reply src/test/skript/tests/syntaxes/effects/EffHealth.sk]: You cannot wait, because anything after the wait will run after the test runner is finished, essentially making it useless.

[the remark this answers is not in the document]
Comment (0XPYEX0) [reply src/test/skript/tests/syntaxes/effects/EffHealth.sk]: > You cannot wait, because anything after the wait will run after the test runner is finished, essentially making it useless.

Yes, so I removed it. But idk how to fix the `mark` problem, bcs the damage is delayed

File: src/test/skript/tests/syntaxes/effects/EffHealth.sk [hunk 28]
```diff
@@ -1,23 +1,31 @@
...
+    heal {_m} by 99
+    assert health of {_m} is 5 with "heal cow failed"
+
+    set {_v::*} to "%minecraft version%" split by "."
+    set {_result} to "%{_v::2}%%{_v::3} ? 0%" parsed as number
+    parse if {_result} >= 204:
+        send "1.20.4+ and test damage cause" to console
+        damage {_m} by 2 hearts with damage cause fall
+        assert health of {_m} is 3 with "damage cow failed"
+        send "Check passed" to console
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/effects/EffHealth.sk:30]:
```suggestion
    if running minecraft "1.20.4":
        damage {_m} by 2 hearts with damage cause fall
        assert health of {_m} is 3 with "damage cow failed"
```

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/effects/EffHealth.sk (hunk 28 above)
Comment (0XPYEX0) [reply src/test/skript/tests/syntaxes/effects/EffHealth.sk:30]: But what about 1.21 and above？

Comment (sovdeeth) [reply src/test/skript/tests/syntaxes/effects/EffHealth.sk:30]: It's a >= check

**Review by sovdeeth — APPROVED** [review]

**Review by cheeezburga — APPROVED** [review]

**Comment by 0XPYEX0** [comment]
Actually, i don't think the test codes are necessary
I cannot solve the `mark` problem, causing the failure of test
The method just marked the entity to be damaged, so assert error

**Comment by Moderocky** [comment]
It looks like the creator needs help with fixing the tests for this. If somebody would be willing to assist, that would be good.

File: src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java [hunk 29]
```diff
@@ -29,108 +37,111 @@ public class HealthUtils {
...
 	/**
 	 * Set the max health an entity can have
-	 * @param e Entity to set max health for
+	 *
+	 * @param damageable      Entity to set max health for
 	 * @param health How many hearts the entity can have at most
 	 */
-	public static void setMaxHealth(Damageable e, double health) {
+	public static void setMaxHealth(Damageable damageable, double health) {
 		AttributeInstance attributeInstance = ((Attributable) e).getAttribute(MAX_HEALTH);
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java:79]:
```suggestion
		AttributeInstance attributeInstance = ((Attributable) damageable).getAttribute(MAX_HEALTH);
```

**Comment by Moderocky** [comment]
So I've investigated this and solved the test issue: these new experimental methods process more 'natural' damage, which goes through the damage logic, triggering damage mitigation, invincibility frames, etc.

I've fixed the test by creating a _new_ entity for the damage source tests, but this might become a problem in future since the behaviour is quite unpredictable.
